src/algebra/algebra/non_unital.lean

```diff
@@ -0,0 +1,163 @@
+  a * (t • b) = t • (a * b) :=
+non_unital_non_assoc_algebra.mul_smul_comm' t a b
+
+/-- The algebra multiplication as a bilinear map. -/
+def mul_as_bilinear {R : Type u₁} {A : Type u₂}
+  [comm_semiring R] [non_unital_non_assoc_semiring A] [non_unital_non_assoc_algebra R A] :
+  A →ₗ[R] A →ₗ[R] A :=
+linear_map.mk₂ R (*) add_mul smul_mul_assoc mul_add mul_smul_comm
+
+/-- If the underlying `non_unital_non_assoc_semiring` is actually a `semiring` we have an
+`algebra`. -/
+def to_algebra {R : Type u₁} {A : Type u₂}
+  [comm_semiring R] [semiring A] [non_unital_non_assoc_algebra R A] : algebra R A :=
+algebra.of_semimodule smul_mul_assoc mul_smul_comm
+
+end non_unital_non_assoc_algebra
+
+variables [non_unital_non_assoc_semiring A] [non_unital_non_assoc_algebra R A]
+variables [non_unital_non_assoc_semiring B] [non_unital_non_assoc_algebra R B]
+variables [non_unital_non_assoc_semiring C] [non_unital_non_assoc_algebra R C]
+
+set_option old_structure_cmd true
+
+/-- A morphism of non-associative algebras. -/
+structure non_unital_non_assoc_algebra_hom extends A →ₗ[R] B, mul_hom A B
+
+attribute [nolint doc_blame] non_unital_non_assoc_algebra_hom.to_linear_map
+attribute [nolint doc_blame] non_unital_non_assoc_algebra_hom.to_mul_hom
+
+namespace non_unital_non_assoc_algebra_hom
+
+variables {R A B C}
+
+instance : has_coe (non_unital_non_assoc_algebra_hom R A B) (A →ₗ[R] B) := ⟨to_linear_map⟩
+
+instance : has_coe (non_unital_non_assoc_algebra_hom R A B) (mul_hom A B) := ⟨to_mul_hom⟩
+
+/-- see Note [function coercion] -/
+instance : has_coe_to_fun (non_unital_non_assoc_algebra_hom R A B) := ⟨_, to_fun⟩
+
+initialize_simps_projections non_unital_non_assoc_algebra_hom (to_fun → apply)
+
+@[simp] lemma coe_mk (f : A → B) (h₁ h₂ h₃) :
+  ((⟨f, h₁, h₂, h₃⟩ : non_unital_non_assoc_algebra_hom R A B) : A → B) = f := rfl
+
+@[simp, norm_cast] lemma coe_to_linear_map (f : non_unital_non_assoc_algebra_hom R A B) :
+  ((f : A →ₗ[R] B) : A → B) = f :=
+rfl
+
+@[simp, norm_cast] lemma coe_to_mul_hom (f : non_unital_non_assoc_algebra_hom R A B) :
+  ((f : mul_hom A B) : A → B) = f :=
+rfl
+
+@[simp] lemma map_smul (f : non_unital_non_assoc_algebra_hom R A B) (c : R) (x : A) :
+  f (c • x) = c • f x :=
+linear_map.map_smul (f : A →ₗ[R] B) c x
+
+@[simp] lemma map_add (f : non_unital_non_assoc_algebra_hom R A B) (x y : A) :
+  f (x + y) = (f x) + (f y) :=
+linear_map.map_add (f : A →ₗ[R] B) x y
+
+@[simp] lemma map_mul (f : non_unital_non_assoc_algebra_hom R A B) (x y : A) :
+  f (x * y) = (f x) * (f y) :=
+mul_hom.map_mul (f : mul_hom A B) x y
+
+@[simp] lemma map_zero (f : non_unital_non_assoc_algebra_hom R A B) : f 0 = 0 :=
+(f : A →ₗ[R] B).map_zero
+
+instance : has_zero (non_unital_non_assoc_algebra_hom R A B) :=
+⟨{ map_mul' := by simp,
+   ..(0 : A →ₗ[R] B)}⟩
+
+instance : inhabited (non_unital_non_assoc_algebra_hom R A B) := ⟨0⟩
+
+lemma coe_injective :
+  function.injective (λ f : non_unital_non_assoc_algebra_hom R A B, show A → B, from f) :=
```

Comment:
This statement leads to some ugly non-delta reduced expressions when applying it, as I recall. Using `@function.injective _ _ coe_fn` with appropriate underscore values resolves that

Reply:
Thanks, noted!